ext/vernier/vernier.cc

```diff
@@ -74,6 +74,34 @@ static const char *gvl_event_name(rb_event_flag_t event) {
     return "no-event";
 }
 
+struct Frame {
+    VALUE frame;
+    VALUE file;
+    int line;
+    
+    Frame() : frame(Qnil), file(Qnil), line(-1) { }
+    
+    Frame(VALUE frame, int line) :
+        frame(frame), 
+        file(frame == 0 ? Qnil : get_frame_file(frame)), 
+        line(line) { }
+    
+    private:
+        static VALUE get_frame_file(VALUE frame) {
+            VALUE file = rb_profile_frame_absolute_path(frame);
+            if (NIL_P(file)) file = rb_profile_frame_path(frame);
+            return file;
+        }
+};
+
+bool operator==(const Frame& lhs, const Frame& rhs) noexcept {
+    return lhs.frame == rhs.frame && lhs.file == rhs.file && lhs.line == rhs.line;
+}
```

Comment:
Including `file` in this comparison is the core of the fix. See https://github.com/jhawthorn/vernier/commit/a2660a7c96cc8425f507472527a574a34fdc7f21 for some context.